### ISSUE TITLE
Feature: Add Happy Eyeballs (RFC 6555) feedback mechanism and sorting

### ISSUE THREAD

**Issue by bradh352** [c-ares/c-ares#609]
ares_getaddrinfo() can return multiple AAAA and A records for a single query.  Happy eyeballs (RFC 6555) defines how to sort these returned addresses based on both the address class and prior address / port connection success/fail in order to quickly connect to the remote service.

Part of a happy eyeballs implementation requires tracking of successful connections in order to influence the future sorting of possible hosts, so a feedback mechanism is required in order to influence future sort orders.  There should be multiple possible values for the feedback, listed in sort order:
1. Successful connection
2. Not tried (default status)
3. Attempted (but don't know status, likely parallel attempt and another succeeded)
4. Timeout
5. Fail

ares_getaddrinfo() already sorts address based on system routing priorities, so the original order returned needs to be part of the sort algorithm.

The general algorithm in the sort will be to first sort by feedback, then if feedback is equal, alternate between ipv6 and ipv4 addresses in ascending original index order.

In order to successfully use this, the ares_getaddrinfo() _**must**_ be provided a port as a hint, and the caller _**must**_ provide feedback via the provided feedback mechanism on each connection success/fail.

The caller also _should_ follow the other Happy Eyeballs recommended practices such as starting multiple connections in parallel after a small delay (e.g. 100ms) before trying the next server quickly be able to locate the best server without waiting on a timeout from each server.

**Comment by bagder** [comment]
Is anyone actually asking for and intending to use a mechanism like this?

The tricky part about storing and caching a lot of info about IP addresses to rely on later is that assumes that the situation for them remains. It requires that they can figure out when the network changes so that the caches can be flushed. That's difficult and error-prone in many cases.

So, how much value will this provide and to whom? I won't stop anyone from working on this, I just think it smells like risking being a lot of work for very little gain.

In curl we've done happy eyeballs correctly for a long time already and I don't see how we will use this API.

**Comment by bradh352** [comment]
I'd probably replace the solution I use in mstdlib (https://github.com/Monetra/mstdlib) with the native c-ares one if it existed.  It might also help spur more adoption of happy eyeballs in other products if people didn't have to reimplement the wheel each time.  

Its just a thought though, its a fairly low priority item, just thought I'd start filing tickets on "ideas" like this one.  

And as far as system changes are concerned, yeah, I was going to file a ticket on that too :)
